readinglist/errors.py

```diff
@@ -0,0 +1,86 @@
+from .utils import json
+from pyramid.config import global_registries
+from pyramid.httpexceptions import (
+    HTTPServiceUnavailable as PyramidHTTPServiceUnavailable, HTTPBadRequest
+)
+from readinglist.utils import Enum
+
+ERRORS = Enum(
+    MISSING_AUTH_TOKEN=104,
+    INVALID_AUTH_TOKEN=105,
+    BADJSON=106,
+    INVALID_PARAMETERS=107,
+    MISSING_PARAMETERS=108,
+    INVALID_POSTED_DATA=109,
+    INVALID_RESOURCE_ID=110,
+    MISSING_RESOURCE=111,
+    FORBIDDEN=121,
+    REQUEST_TOO_LARGE=113,
+    CLIENT_REACHED_CAPACITY=117,
+    UNDEFINED=999,
+    BACKEND=201
+)
+
+
+def get_formatted_error(code, errno, error, message=None, info=None):
+    result = {
+        "code": code,
+        "errno": errno,
+        "error": error
+    }
+
+    if message is not None:
+        result['message'] = message
+
+    if info is not None:
+        result['info'] = info
+
+    return json.dumps(result)
+
+
+class HTTPServiceUnavailable(PyramidHTTPServiceUnavailable):
+    """Return an HTTPServiceUnavailable formatted error."""
+
+    def __init__(self, **kwargs):
+        if 'body' not in kwargs:
+            kwargs['body'] = get_formatted_error(
+                503, ERRORS.BACKEND, "Service unavailable",
+                "Service unavailable due to high load, please retry later.")
+
+        if 'content_type' not in kwargs:
+            kwargs['content_type'] = 'application/json'
+
+        if 'headers' not in kwargs:
+            kwargs['headers'] = []
+
+        settings = global_registries.last.settings
+        retry_after = settings.get(
+            'readinglist.retry_after', "30").encode("utf-8")
+        kwargs['headers'].append(
+            ("Retry-After", retry_after)
+        )
+
+        super(HTTPServiceUnavailable, self).__init__(**kwargs)
+
+
+def json_error(errors):
+    """Returns an HTTPError with the given status and message.
+
+    The HTTP error content type is "application/json"
+    """
+    sorted_errors = sorted(errors, key=lambda x: "%s" % x['name'])
```

Comment:
I don't get why we need to do the "%s" trick here. wouldn't `sorted(errors, key=lambda x: x['name'])` be enough? or if we really need to convert to a string, then use `str`.